perun-core/src/main/java/cz/metacentrum/perun/core/impl/AttributesManagerImpl.java

```diff
@@ -3440,19 +3481,57 @@ public void deleteAllAttributeAuthz(PerunSession sess, AttributeDefinition attri
 			return hashMap;
 		}
 
-	    try {
-			return jdbc.query("SELECT " + getAttributeMappingSelectQuery("mem") + ", members.id FROM attr_names " +
-							"JOIN service_required_attrs ON attr_names.id=service_required_attrs.attr_id AND service_required_attrs.service_id=? " +
-							"JOIN members ON " + BeansUtils.prepareInSQLClause("members.id", members) +
-							"LEFT JOIN member_attr_values mem ON attr_names.id=mem.attr_id " +
-							"AND mem.member_id=members.id WHERE namespace IN (?,?,?,?)",
-					new MemberAttributeExtractor(sess, this, members), service.getId(),
-					AttributesManager.NS_MEMBER_ATTR_CORE, AttributesManager.NS_MEMBER_ATTR_DEF, AttributesManager.NS_MEMBER_ATTR_OPT, AttributesManager.NS_MEMBER_ATTR_VIRT);
-		} catch (EmptyResultDataAccessException ex) {
-			return new HashMap<>();
-		} catch (RuntimeException ex) {
-			throw new InternalErrorException(ex);
-		}
+		try {
+			if (Compatibility.isOracle()) {
+				return jdbc.execute("SELECT " + getAttributeMappingSelectQuery("mem") + ", members.id FROM attr_names " +
+					"JOIN service_required_attrs ON attr_names.id=service_required_attrs.attr_id AND service_required_attrs.service_id=? " +
+					"JOIN members ON members.id in (SELECT * FROM TABLE(?)) " +
+					"LEFT JOIN member_attr_values mem ON attr_names.id=mem.attr_id " +
+					"AND mem.member_id=members.id WHERE namespace IN (?,?,?,?)", (PreparedStatementCallback<HashMap<Member, List<Attribute>>>) preparedStatement -> {
+					Array sqlArray = ((PerunBlImpl)perun).getDatabaseManagerBl().prepareOracleArrayOfNumbers(members, preparedStatement);
+					preparedStatement.setInt(1, service.getId());
+					preparedStatement.setArray(2, sqlArray);
+					preparedStatement.setNString(3, AttributesManager.NS_MEMBER_ATTR_CORE);
+					preparedStatement.setNString(4, AttributesManager.NS_MEMBER_ATTR_DEF);
+					preparedStatement.setNString(5, AttributesManager.NS_MEMBER_ATTR_OPT);
+					preparedStatement.setNString(6, AttributesManager.NS_MEMBER_ATTR_VIRT);
+					MemberAttributeExtractor memberAttributeExtractor = new MemberAttributeExtractor(sess, this, members);
+					return memberAttributeExtractor.extractData(preparedStatement.executeQuery());
+				});
+			}
+		} catch (InternalErrorRuntimeException ex) {
+			//This is possible if oracle sql array can't be created successfully (for example if oracle driver is missing)
+			//If happen, log it as error and continue using the standard way to obtain objects by IN clause
+			log.error("Unable to use Oracle SQL Array to obtain objects from database! Standard way to obtain these objects will be used instead.", ex);
+		}
+
+		HashMap<Member, List<Attribute>> allReturnedAttrs = new HashMap<>();
+		int from = 0;
+		int to = MAX_SIZE_OF_BULK_IN_SQL;
+
+		//Do not call one query for more than MAX_SIZE_OF_BULK_IN_SQL values in "SQL IN CLAUSE"
+		//Reason: SQL error in DB for too much records in one SQL query
+		do {
```

Comment:
again, for PostgreSQL and HSQL, the same trick with array can be used instead of generating IN clause with thousands of constants